api/http/management_filetransfer.go

```diff
@@ -0,0 +1,760 @@
+	numberOfChunks := 0
+	for {
+		select {
+		case wsMessage := <-msgChan:
+			// reset the timeout ticket
+			timeout.Reset(fileTransferTimeout)
+			// process the message
+			err := h.downloadFileResponseProcessMessage(c, params, request,
+				wsMessage, deviceTopic, &latestOffset, &numberOfChunks,
+				&responseHeaderSent, ticker)
+			if err == io.EOF {
+				return
+			} else if err != nil {
+				responseError = err
+				return
+			}
+		// send a Ping message to keep the session alive
+		case <-ticker.C:
+			h.publishFileTransferPing(params.SessionID, deviceTopic)
+
+		// no message after timeout expired, stop here
+		case <-timeout.C:
+			responseError = errFileTranserTimeout
+			return
+		}
+	}
+}
+
+func (h ManagementController) downloadFileResponseProcessMessage(c *gin.Context,
+	params *fileTransferParams, request *model.DownloadFileRequest, wsMessage *natsio.Msg,
+	deviceTopic string, latestOffset *int64, numberOfChunks *int, responseHeaderSent *bool,
+	ticker *time.Ticker) error {
+	msg, msgBody, err := h.decodeFileTransferProtoMessage(wsMessage.Data)
+	if err != nil {
+		return err
+	}
+
+	// process incoming messages from the device by type
+	switch msg.Header.MsgType {
+
+	// error message, stop here
+	case wsft.MessageTypeError:
+		errorMsg := msgBody.(*wsft.Error)
+		if *errorMsg.MessageType == wsft.MessageTypeStat {
+			return errors.Wrap(errors.New(*errorMsg.Error),
+				errFileTranserFailed.Error())
+		} else {
+			return errors.New(*errorMsg.Error)
+		}
+
+	// file stat response, if okay, let's get the file
+	case wsft.MessageTypeFileInfo:
+		req := wsft.GetFile{
+			Path: request.Path,
+		}
+		if err := h.publishFileTransferProtoMessage(params.SessionID,
+			params.UserID, deviceTopic, wsft.MessageTypeGet,
+			req, 0); err != nil {
+			return err
+		}
+
+		fileInfo := msgBody.(*wsft.FileInfo)
+		writeHeaders(c, fileInfo)
+		*responseHeaderSent = true
+
+	// file data chunk
+	case wsft.MessageTypeChunk:
+		if msg.Body == nil {
+			if err := h.publishFileTransferProtoMessage(
+				params.SessionID, params.UserID, deviceTopic,
+				wsft.MessageTypeACK, nil,
+				*latestOffset); err != nil {
+				return err
+			}
+			return io.EOF
+		}
+
+		// verify the offset property
+		propOffset, _ := msg.Header.Properties[PropertyOffset].(int64)
+		if propOffset != *latestOffset {
+			return errors.Wrap(errFileTranserFailed,
+				"wrong offset received")
+		}
+		*latestOffset += int64(len(msg.Body))
+
+		_, err := c.Writer.Write(msg.Body)
+		if err != nil {
+			return err
+		}
+
+		(*numberOfChunks)++
+		if *numberOfChunks >= ackSlidingWindowSend {
+			if err := h.publishFileTransferProtoMessage(
+				params.SessionID, params.UserID, deviceTopic,
+				wsft.MessageTypeACK, nil,
+				*latestOffset); err != nil {
+				return err
+			}
+			*numberOfChunks = 0
+		}
+
+	case ws.MessageTypePing:
+		if err := h.publishFileTransferProtoMessage(
+			params.SessionID, params.UserID, deviceTopic,
+			ws.MessageTypePong, nil,
+			-1); err != nil {
```

Comment:
This message has wrong protocol ID (`ws.ProtoTypeFIleTransfer` - should have `ws.ProtoTypeSessionControl`) and you'll get an error message in response.

Reply:
I don't understand the "and you'll get an error message in response".

Reply:
Ohh, sorry. Reading it again I see I'm very imprecise; what I meant is since the message has wrong `proto` header, it will be routed to the handler for file transfer messages which in turn will return an error saying it don't understand the message type.